File: src/main/java/com/now/nowbot/config/AsyncSetting.java
package com.now.nowbot.config;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.task.AsyncTaskExecutor;
import org.springframework.core.task.TaskExecutor;
import org.springframework.scheduling.annotation.AsyncConfigurer;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
import org.springframework.stereotype.Component;

import java.util.concurrent.Executor;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;

/***
 * 线程池配置
 */
@Component
@EnableAsync
@Configuration
public class AsyncSetting implements AsyncConfigurer {
    private static final Logger log = LoggerFactory.getLogger(AsyncSetting.class);
    public static final ThreadFactory V_THREAD_FACORY = Thread.ofVirtual().name("Bot").factory();
    static private final ThreadPoolTaskExecutor threadPool;

    static {
        threadPool = new ThreadPoolTaskExecutor();
        threadPool.setThreadFactory(V_THREAD_FACORY);
        threadPool.setCorePoolSize(1000);
        threadPool.setMaxPoolSize(Integer.MAX_VALUE);
        threadPool.setKeepAliveSeconds(5);
        threadPool.setQueueCapacity(Runtime.getRuntime().availableProcessors() * 10);
        threadPool.setRejectedExecutionHandler(new ThreadPoolExecutor.DiscardOldestPolicy());
        threadPool.setWaitForTasksToCompleteOnShutdown(true);
        threadPool.setAwaitTerminationSeconds(5);
        threadPool.initialize();
    }

    @Bean("botAsyncExecutor")
    public ThreadPoolTaskExecutor executor() {
        return threadPool;
    }

    @Bean(name = {"mainExecutor", "shiroTaskExecutor"})
    public Executor getAsyncExecutor() {
        return AsyncSetting.threadPool;
    }

    /**
     * 自定义异常处理
     */
    @Override
    public AsyncUncaughtExceptionHandler getAsyncUncaughtExceptionHandler() {
        return (throwable, method, obj) -> {
            log.info("Method data - {}", method.getName(), throwable);
            for (Object param : obj) {
                log.info("Parameter value - {}", param.toString());
            }
        };
    }

    @Bean(name = {"applicationTaskExecutor", "taskExecutor"})
    public TaskExecutor getTaskExecutor(){
        return AsyncSetting.threadPool;
    }

    @Bean
    public AsyncTaskExecutor AstncConf() {
        return AsyncSetting.threadPool;
    }
}

File: src/main/java/com/now/nowbot/config/AsyncSetting.kt
package com.now.nowbot.config

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.annotation.AsyncConfigurer
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.util.concurrent.Executor
import java.util.concurrent.ThreadFactory
import java.util.concurrent.ThreadPoolExecutor

@Configuration
@EnableAsync
class AsyncSetting : AsyncConfigurer {

    // 将 threadPool 移动到 Spring Bean 管理下，不要放在 companion object 里
    @Bean(name = ["botAsyncExecutor", "mainExecutor", "shiroTaskExecutor", "applicationTaskExecutor", "taskExecutor", "kotlinTaskExecutor"])
    fun threadPoolTaskExecutor(): ThreadPoolTaskExecutor {
        val executor = ThreadPoolTaskExecutor()

        // 关键：使用虚拟线程工厂
        executor.setThreadFactory(threadFactory)

        executor.corePoolSize = 100
        executor.maxPoolSize = 2000 // 虚拟线程不需要太大的池管理，其实可以直接用虚拟线程执行器
        executor.queueCapacity = 500

        // 允许优雅停机
        executor.setWaitForTasksToCompleteOnShutdown(true)
        executor.setAwaitTerminationSeconds(10) // 给10秒收尾时间

        executor.setRejectedExecutionHandler(ThreadPoolExecutor.CallerRunsPolicy())
        executor.initialize()
        return executor
    }

    // 实现接口方法，直接注入上面定义的 Bean
    override fun getAsyncExecutor(): Executor {
        return threadPoolTaskExecutor()
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(AsyncSetting::class.java)

        internal class ThreadFactoryBox(private val threadFactory: ThreadFactory) : ThreadFactory {
            override fun newThread(r: Runnable): Thread? {
                return threadFactory.newThread {
                    try {
                        r.run()
                    } catch (e: Throwable) {
                        log.error("thread throw: ", e)
                    }
                }
            }
        }

        // 仅保留 Factory
        private val virtualThreadFactory: ThreadFactory = Thread.ofVirtual().name("Bot-", 0).factory()
        val threadFactory: ThreadFactory = ThreadFactoryBox(virtualThreadFactory)
    }
}